rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1421 @@
+            },
+            // Decimal
+            TestCase {
+                source_arrays: vec![
+                    Arc::new(Decimal128Array::from(vec![53, 42])),
+                    Arc::new(Decimal128Array::from(vec![68, 32])),
+                ],
+                expected_min: ScalarValue::try_new_decimal128(32, 38, 10).unwrap(),
+                expected_max: ScalarValue::try_new_decimal128(68, 38, 10).unwrap(),
+            },
+        ];
+
+        for case in cases {
+            let array_refs = case.source_arrays.iter().collect::<Vec<_>>();
+            let stats = collect_statistics(&array_refs);
+            assert_eq!(
+                stats,
+                StatisticsRow {
+                    null_count: ScalarValue::Int64(Some(0)),
+                    min_value: case.expected_min,
+                    max_value: case.expected_max,
+                },
+                "Statistics are wrong for input data: {:?}",
+                case.source_arrays
+            );
+        }
+    }
+
+    #[test]
+    fn test_collect_float_stats() {
+        // NaN values are ignored in statistics
+        let arrays: Vec<ArrayRef> = vec![
+            Arc::new(Float32Array::from(vec![4.0f32, 3.0, std::f32::NAN, 2.0])),
+            Arc::new(Float32Array::from(vec![-10.0f32, 3.0, 5.0, std::f32::NAN])),
+        ];
+        let array_refs = arrays.iter().collect::<Vec<_>>();
+        let stats = collect_statistics(&array_refs);
+        assert_eq!(
+            stats,
+            StatisticsRow {
+                null_count: ScalarValue::from(0_i64),
+                min_value: ScalarValue::from(-10.0_f32),
+                max_value: ScalarValue::from(5.0_f32),
+            }
+        );
+
+        // (Negative) Infinity can be min or max.
+        let arrays: Vec<ArrayRef> = vec![Arc::new(Float64Array::from(vec![
+            4.0f64,
+            std::f64::INFINITY,
+            std::f64::NEG_INFINITY,
+        ]))];
+        let array_refs = arrays.iter().collect::<Vec<_>>();
+        let stats = collect_statistics(&array_refs);
+        assert_eq!(
+            stats,
+            StatisticsRow {
+                null_count: ScalarValue::from(0_i64),
+                min_value: ScalarValue::from(std::f64::NEG_INFINITY),
+                max_value: ScalarValue::from(std::f64::INFINITY),
+            }
+        );
+
+        // Max value for zero is always positive, min value for zero is always negative
+        let arrays: Vec<ArrayRef> = vec![Arc::new(Float32Array::from(vec![-0.0, 0.0]))];
+        let array_refs = arrays.iter().collect::<Vec<_>>();
+        let stats = collect_statistics(&array_refs);
+        assert_eq!(
+            stats,
+            StatisticsRow {
+                null_count: ScalarValue::from(0_i64),
+                min_value: ScalarValue::from(-0.0_f32),
+                max_value: ScalarValue::from(0.0_f32),
+            }
+        );
+    }
+
+    #[test]
+    fn test_collect_binary_stats() {
```

Comment:
Could we also test that values that are *exactly* the length of the limit don't have their max incremented. That is, it still preserved the invariant that an array of just value `V` has stats `{min: V, max: V}`?

Reply:
Added tests.